### ISSUE TITLE
Allow custom build-in page types for forked instance for migrating from custom repo pages

### ISSUE THREAD

**Issue by dzhuang** [inducer/relate#475]
1. Allow custom build-in page types
1. Backward compatibility (after deadline was imposed, no new fetch is allowed, while allow previous validated custom repo pages to run,). 

Any suggestions? @inducer

**Comment by inducer** [comment]
What do you mean by 'custom'?

How grumpy would you be if I said that it's easy enough to maintain a fork that keeps allowing this?

**Comment by dzhuang** [comment]
Maintaining a fork is ok for me (and that's what I were doing). I'm seek a way to do that by something settings.

**Comment by dzhuang** [comment]
For example, in settings:

```Python
RELATE_CUSTOM_BUILT_IN_PAGE_TYPES = (
    "my_custom_pages.custom_page1",
    "my_custom_pages.custom_page2")
```

Then in 'course.content.get_flow_page_class`, we use django's `import_string` to import the custom class. However, the problem is, 'my_custom_pages' should also be mained in the fork, because it need to subclass RELATE's built-in pages.

**Comment by inducer** [comment]
Not sure I understand this concern. IIRC, dotted-name import of page types is already tried, and all importable page types are considered "safe". Are you suggesting wrapping a whitelist around that?

Also, these page types don't need to be in Relate itself. You can just install them into the same virtualenv as, or add them to `sys.path` of the app server.

Or am I totally misunderstanding you?

**Comment by dzhuang** [comment]
> Are you suggesting wrapping a whitelist around that?

You're right. 

> You can just install them into the same virtualenv as..

If this route, in terms of developing, I need first install RELATE as a dependancy, am I right?

**Comment by inducer** [comment]
> Whitelist

Sure, it's not impossible that we might import something we don't actually want to import that way. OK with me.

> If this route, in terms of developing, I need first install RELATE as a dependancy, am I right?

Not necessarily. Since there's no 'compile time', you can just install that independently.

**Comment by dzhuang** [comment]
> OK with me.
> You can just install that independently.

Thanks.

**Comment by dzhuang** [comment]
What's more, I have a feeling that `course.validation.check_for_page_type_changes` should be tolerable for when custom pages were switched from custom repo page types to what I called custom built-in page types.

**Comment by dzhuang** [comment]
(i.e., `reop:my_custom_page_type1` is considered the same with `my_custom_pages.custom_page_type1`)

**Comment by inducer** [comment]
Yeah, that's a good point. A way to do that would be for the class to contain an attribute

```
page_type_name_aliases = [
    "repo:stuff.some_classname",
    "repo:stuff.another_classname",
    ]
```
or
```python
page_type_name_aliases = [
   ("cs101", "repo:stuff.some_classname"),  # course-specific match
   (None, "repo:stuff.another_classname"),
    ]
```

Or you could specify the mapping as part of the configuration. Not sure I have a strong preference.

**Comment by dzhuang** [comment]
Thanks. I think a mapping in setting is ok.
